### ISSUE TITLE
Automated Deployment of Magma in Kubernetes Cluster

### ISSUE THREAD

**Issue by jordanvrtanoski** [magma/magma#15357]
### Problem

Currently, Magma can be deployed within a K8s cluster, however the cluster nodes should be manually prepared with the OVS as a part of the node infrastructure.

### Solution

Our proposal is based on the standard Kubernetes architecture and will allow deployment of Magma on the nodes without the requirement to prepare the nodes in a special way before the deployment. The limitation of the supported kernel version will remain (due to the current version of the GTP and OVS kernel modules), however in the future we can have a proposal to address this part as well.

### Proposal 

This is a proposal for adding support for Kubernetes in Magma. The main challenge is to enable communication between the eNodeB and the Magma gateway using GTP tunnels. To achieve this, we propose to use a pod as DeamonSet for deploying vport-gtp custom kernel module, which allows OVS to handle GTP packets. Additionally, we need to deploy another pod for OVS, which will create bridges and flows for routing the traffic between the eNodeB and the gateway. This approach has the advantage of being scalable, flexible and compatible with different Kubernetes environments. 

Issue #[14833](https://github.com/magma/magma/issues/14833) added support for Kubernetes However, but the custom vport-gtp module build and its installation is not yet automated.

![daemonset-explained](https://github.com/magma/magma/assets/1470453/678b37b2-4036-48bb-875a-2b3700f67042)
Image credits: https://devopscube.com/kubernetes-daemonset/ 

### Activities 

- The high-level list of activities. 
- Setup Kubernetes Cluster for development 
- Deploy and test Magma with current helm charts and apply any fixes if necessary. 
- Create the kernel module build pod under a DeamonSet. 
- Create the OVS pod and integrate it with the rest of magma components. 
- Setup a Kubernetes cluster ready for production environment. 
- Deploy Magma in cluster 
- Test Magma LTE  
- Test Magma 5GNR 
- Integrate with DevOps infrastructure 

### Required effort and budget

Activity | Duration (week/person) | Cost in USD
-- | -- | --:
Setup Kubernetes Cluster for development | 0.5 | 1.000
Deploy and test Magma with current helm charts and apply any fixes if necessary. | 0.5 | 1.000
Create the kernel module build pod under a DeamonSet. | 2 | 4.000
Create the OVS pod and integrate it with the rest of magma components. | 1 | 2.000
Setup a Kubernetes cluster ready for production environment and deploy Magma | 1 | 2.000
Test Magma LTE | 0.5 | 1.000
Test Magma 5GNR | 0.5 | 1.000
Integrate with DevOps infrastructure | 2 | 4.000
Total | 8 | 16.000

### Non-goals

- Reengineering the GTP kernel module
- Load and balancing tests

**Comment by panyogesh** [comment]
+1

**Comment by lucasgonze** [comment]
Lucas: +1

Raphael: +1

The motion is carried.

**Comment by lucasgonze** [comment]
Blocked on merge above because of failures in integration tests for domain proxy. These are test failures, not feature breakage. Fixes are WIP. Yogesh will review to see if Veltris can offer insights.

**Comment by lucasgonze** [comment]
All tests are passing except domain proxy. So this is blocked on https://github.com/magma/magma/issues/15573

**Comment by lucasgonze** [comment]
Final blocker expected to be cleared this week - green in CI.
